syntax = "proto3"; 
package chatnow; 
option cc_generic_services = true; 
 
//用户信息结构 
message UserInfo { 
    string user_id = 1;//用户 ID 
    string nickname = 2;//昵称 
    string description = 3;//个人签名/描述 
    string mail = 4; //绑定手机号 
    bytes avatar = 5;//头像照片，文件内容使用二进制 
} 

/* brief: 聊天会话状态 */
enum ChatSessionStatus {
    NORMAL_SESSION    = 0;
    ARCHIVED_SESSION  = 1;
    DISMISSED_SESSION = 2;
}

//聊天会话成员权限（群聊用）
enum ChatSessionMemberRole {
    NORMAL_ROLE = 0;
    ADMIN_ROLE  = 1;
    OWNER_ROLE  = 2;
}

// 会话列表中，'我'在该会话下的状态摘要
message MySessionState {
    bool is_muted = 1;                 // 是否免打扰
    bool is_visible = 2;
    bool pin_time = 3;                // 是否置顶（由 pin_time != null 推导）
}

//聊天会话信息（加入前)
message ChatSessionBrief {
    string chat_session_id = 1; //会话 ID 
    string chat_session_name = 2;//会话名称 git
    optional string avatar = 3; //原来是bytes
    uint32 chat_session_type = 4;
    int64 create_time = 5;   //会话创建时间
    uint32 member_count = 6; //会话人员数量
}

//聊天会话信息（加入后）
message ChatSessionInfo { 
    //群聊会话不需要设置，单聊会话设置为对方用户 ID 
    optional string single_chat_friend_id = 1; 
    string chat_session_id = 2; //会话 ID 
    string chat_session_name = 3;//会话名称 git 
    //会话上一条消息，新建的会话没有最新消息 
    optional MessageInfo prev_message = 4; 
    //会话头像
    optional string avatar = 5; //原来是bytes
    //============ V2.0 ================= 
    uint32 chat_session_type = 6;
    int64 create_time = 7;   //会话创建时间
    uint32 member_count = 8; //会话人员数量
    ChatSessionStatus status = 9; //会话状态
    oneof member_info {
        MySessionState member_info_brief = 10; 
        ChatSessionMemberInfo member_info_detail = 11;
    }
}

//聊天会话成员信息
message ChatSessionMemberInfo { 
    //所在会话ID
    string chat_session_id = 1;
    //自身ID
    string user_id = 2;
    optional uint64 last_message_id = 3;
    bool is_muted = 4;
    bool is_visible = 5;
    optional int64 pin_time = 6;
    ChatSessionMemberRole role = 7;
    int64 join_time = 8;
} 

message ChatSessionMemberItem {
    UserInfo user_info = 1;                 // 用户基础信息（来自用户服务）
    ChatSessionMemberRole role = 2;         // 群内身份
    int64 join_time = 3;                    // 加入时间
}

// 成员-角色配对结构体
message MemberRolePair {
    string user_id = 1;
    ChatSessionMemberRole role = 2;
}

//消息类型 
enum MessageType { 
    STRING = 0; 
    IMAGE = 1; 
    FILE = 2; 
    SPEECH = 3; 
} 
message StringMessageInfo { 
    string content = 1;//文字聊天内容 
} 
message ImageMessageInfo { 
    //图片文件 id,客户端发送的时候不用设置，由 transmit 服务器进行设置后交给 storage 的时候设置 
    optional string file_id = 1; 
    //图片数据，在 ES 中存储消息的时候只要 id 不要文件数据, 服务端转发的时候需要原样转发 
    optional bytes image_content = 2; 
} 
message FileMessageInfo { 
    optional string file_id = 1;//文件 id,客户端发送的时候不用设置 
    optional int64 file_size = 2;//文件大小 
    optional string file_name = 3;//文件名称 
    //文件数据，在 ES 中存储消息的时候只要 id 和元信息，不要文件数据, 服务端转发的时候也不需要填充 
    optional bytes file_contents = 4; 
} 
message SpeechMessageInfo { 
    //语音文件 id,客户端发送的时候不用设置 
    optional string file_id = 1; 
    //文件数据，在 ES 中存储消息的时候只要 id 不要文件数据, 服务端转发的时候也不需要填充 
    optional bytes file_contents = 2; 
} 
message MessageContent { 
    MessageType message_type = 1; //消息类型 
    oneof msg_content { 
        StringMessageInfo string_message = 2;//文字消息 
        FileMessageInfo file_message = 3;//文件消息 
        SpeechMessageInfo speech_message = 4;//语音消息 
        ImageMessageInfo image_message = 5;//图片消息 
    }; 
} 
//消息结构 
message MessageInfo { 
    int64 message_id = 1;//消息 ID 
    string chat_session_id = 2;//消息所属聊天会话 ID 
    int64 timestamp = 3;//消息产生时间 
    UserInfo sender = 4;//消息发送者信息 
    MessageContent message = 5; 
 } 
 
message FileDownloadData { 
    string file_id = 1; 
    bytes file_content = 2; 
} 
 
message FileUploadData { 
    string file_name = 1; 
    int64 file_size = 2; 
    bytes file_content = 3; 
} 